src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -1159,12 +1159,15 @@ internal virtual void ProcessLoggingEvent(object buildEvent, bool allowThrottlin
         /// </summary>
         internal void WaitForThreadToProcessEvents()
         {
-            // This method may be called in the shutdown submission callback, this callback may be called after the logging service has 
-            // shutdown and nulled out the events we were going to wait on.
-            if (_logMode == LoggerMode.Asynchronous && _loggingQueue != null)
+            lock (_lockObject)
```

Comment:
Is this change related to the fix or does it address another issue?

Reply:
It is another bug. `TerminateLoggingEventQueue()` put LoggingService into state which will either silently reject events coming into its queue while it is waiting for all existing events to be handled.
Also `CreateLoggingEventQueue()` modifies `_loggingQueue` which shall be protected for concurrent access.
However, I will have to modify it, because the way I have fixed it, it could cause different deadlock, when any registered logger callback into LoggingService for example with intent to log something. That seems to me as quite possible.
I'll modify it and let you know.